README.md
## A serial communication interface for R.

| OSX and Linux Status | Windows Status | CRAN Version | Downloads|
|-----------------------|----------------|--------------|---------|
|[![Build Status](https://travis-ci.org/aforren1/serial.svg)](https://travis-ci.org/aforren1/serial)|[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/aforren1/serial?branch=master&svg=true)](https://ci.appveyor.com/project/aforren1/serial)|[![CRAN](http://www.r-pkg.org/badges/version/serial)](http://cran.rstudio.com/package=serial)|[![Downloads](http://cranlogs.r-pkg.org/badges/serial?color=brightgreen)](http://www.r-pkg.org/pkg/serial)|


## Installation

To get the released version from CRAN, type:

    install.packages("serial")

Or to install the development version from Github:

    # install.packages("devtools") # only if devtools not installed
    devtools::install_github("aforren1/serial")

NEWS
serial 1.2
-------------------------------------------------------

NEW FEATURES

* Added `.onAttach()` to attempt to catch older OSX installations with incomplete
    tcltk installations.
* Added `findDevices()` helper function, which should show active serial devices on
    OSX, Linux, and Windows.
* Deprecated `read.serialConnection` and `write.serialConnection` for 
    `read` and `write` generics, and associated "serialConnection" 
    methods. This standardizes the different functions that can act on 
    these objects. Additionally, export a generic version of `write` from the base package.


BUG FIXES AND MINOR IMPROVEMENTS

* Import methods, utils, and tcltk.
* Removed `try()` calls in functions to allow easier error handling by the user.
* Removed the `name` argument to `serialConnection()`, because it was not optional
    as stated by the help file and did not seem to serve any purpose.
* Reorganized *.R files.
* Minor refactoring of existing functions.
* Various spelling and grammatical corrections, and documentation improvements.
* Added '-timeout 5000' to `.Tcl` call in `open.serialConnection`. I had an issue
    with OSX hanging and wanted to see if that was the issue.


tests/testthat.R
library(testthat)
library(serial)

test_check("serial")


tests/testthat/test-access.R
library(serial)
context("Accessibility")
#err_msg1 <- paste("[tcl] couldn\'t open ", '"', "/dev/com1", '":', " no such file or directory.\n", sep = "")
#err_msg2 <- paste("[tcl] can\'t read ", '"',"sdev com1", '":', " no such variable.", sep = "")
con <- serialConnection('com40')
test_that("all functions are accessible", {
  expect_error(open(con))
  expect_error(read_data(con))
  expect_error(write_data(con, 'test'))
  #expect_error(write_data(con, 3)) # errors *and* warnings
  expect_error(read.serialConnection(con))
  expect_error(write.serialConnection(con, 'test'))
  #expect_error(write.serialConnection(con, 3))
  expect_error(close(con))
})

tests/testthat/test-virtual.R
# this test only works on linux, and is dependent 
# on already having a virtual device set up (which I failed to do in travis)
# library(serial)
# context('test virtual device')
# 
# con <- serialConnection('tnt0')
# con2 <- serialConnection('tnt1')
# test_that("Functions work for virtual device", {
#   expect_silent(open(con))
#   expect_silent(open(con2))
#   expect_silent(write_data(con, 'test'))
#   expect_warning(write_data(con, 2))
#   expect_is(read_data(con), 'character') #TODO: actually read *some* input
#   expect_silent(close(con))
#   expect_silent(close(con2))
# })

R/init.R
#' Sets up the interface parameters.
#' 
#' @param port  COM port name, should also work in Linux. Also, virtual COMs are 
#'              supported -- USB should work too.
#' @param mode  communication mode "\code{<BAUD>, <PARITY>, <DATABITS>, <STOPBITS>}"
#' \describe{
#'    \item{\code{BAUD}}{sets the baud rate (bits per second)}
#'    \item{\code{PARITY}}{\emph{n, o, e, m, s} stands for "none", "odd", "even", "mark" and "space"}
#'    \item{\code{DATABITS}}{integer number of data bits. The value can range from 5 to 8}
#'    \item{\code{STOPBITS}}{integer number of stop bits. This can be "1" or "2"}
#'        }       
#'              
#' @param buffering "\code{none}", for RS232 serial interface only; other modes don't work in this case
#' @param newline \code{<BOOL>}, whether a new transmission starts with a newline or not
#'                \describe{
#'                  \item{\code{TRUE} or 1}{send newline-char according to \code{<translation>} before transmitting}
#'                  \item{\code{FALSE} or 0}{no newline}
#'                          }
#' @param handshake determines the type of handshaking the communication
#'                  \describe{
#'                    \item{"\code{none}"}{no handshake is done}
#'                    \item{"\code{rtscts}"}{hardware handshake is enabled}
#'                    \item{"\code{xonxoff}"}{software handshake via extra characters is enabled}
#'                    }
#' 
#' @param eof \code{<CHAR>}, termination char of the datastream. It only makes sense
#'        if \code{<translation>} is 'binary' and the stream is a file
#' @param translation each transmitted string is terminated by the transmission
#'       character. This could be 'lf', 'cr', 'crlf', or 'binary'
#' @return An object of the class "\code{serialConnection}" is returned.
#' @export
serialConnection <- function(port = "com1", mode = "115200, n, 8, 1", 
                             buffering = "none", newline = 0, eof = "",
                             translation = "lf", handshake= "none") {
  
  obj <- as.list(environment())
  class(obj) <- "serialConnection"
  obj
}

#' Function to initialize an serial interface.
#' 
#' This function initializes the serial interface and opens it for later usage. 
#' 
#' @method open serialConnection
#' @aliases open
#' 
#' @param con Object of class \code{serialConnection}.
#' @param ... Currently ignored.
#' @seealso \code{\link{serialConnection}} \code{\link{close}}
#' @examples
#' \dontrun{
#' con <- serialConnection(port = 'com5')
#' open(con)
#' }
#' @import tcltk
#' @export
open.serialConnection <- function(con, ...) {
  ## set platform-dependent path
  os_path <- switch(.Platform$OS.type,
                    windows = "//./",
                    unix = "/dev/")
  ## set connection and variables
  .Tcl(paste("set sdev_", con$port," [open ", os_path, con$port, " r+]", sep = ""))
  ## '//./' defines the Windows path for the COM ports
  ## only in this way it is possible to use virtual ports as well
  ## the unix COM ports are located in '/dev/'
  
  ## set up configuration
  eof <- ifelse(con$eof == "", "", paste(" -eofchar ", con$eof, sep = ""))
  
  .Tcl( paste("fconfigure $sdev_",con$port,
              " -mode ", con$mode,
              " -buffering ", con$buffering,
              " -blocking 0", eof,
              " -translation ", con$translation,
              " -handshake ", con$handshake, sep = ""))
  invisible(NULL)
  ## it seems that -eofchar doesn't work
  ## "buffering none" is recommended, other setings doesn't work to send 
}

#' Function to close a serial interface.
#' 
#' This function closes the corresponding connection.
#' 
#' @method close serialConnection
#' @aliases close
#' @param con Object of class \code{serialConnection}.
#' @param ... Currently ignored.
#' 
#' @examples
#' \dontrun{
#' close(con)
#' }
#' 
#' @seealso \code{\link{serialConnection}} \code{\link{open}}
#' @import tcltk
#' @export
close.serialConnection <- function(con, ...) {
  .Tcl(paste("close $sdev_", con$port, sep = ""))
  invisible(NULL)
}

R/pkgname.R
#' A serial communication interface for R.
#' 
#' This R package provides the functionality to use the serial communication ports
#' "COM" to use RS232/RS422/RS485 functionality of the corresponding hardware.
#' Also, virtual COM ports via USB work, as long as they are mapped to COM[n] (Windows)
#' or tty[n] (Mac/Linux) in the operating system.
#' 
#' @description
#' \describe{
#'  \item{\code{\link{serialConnection}}}{Creates a \code{serialConnection} object.}
#'  \item{\code{\link{open}}}{opens a serial connection.}
#'  \item{\code{\link{close}}}{closes the serial connection.}
#'  \item{\code{\link{read}}}{bytewise read from the interface, as long as the buffer is empty.}
#'  \item{\code{\link{write}}}{writes a string to the serial interface.}
#'  \item{\code{\link{FindDevices}}}{tries to find available serial devices.}
#' }
#' 
#' @examples
#' # For this example, I used the 'null-modem' emulator 'com0com' for Windows
#' # which is available on 'http://com0com.sourceforge.net/'.
#' # Here, the pair of COM ports is 'CNCA0' <-> 'CNCB0'.
#' 
#' # For Linux, 'tty0tty' and installation instructions are available on
#' # 'https://github.com/freemed/tty0tty'. The pair of COM ports is 'tnt0' <-> 'tnt1'.
#' 
#' # Test the functionality:
#' # ======================
#' #
#' # First: Install the virtual null-modem connection like
#' #        com0com (win) or tty0tty (linux)
#' #        Hint: Some unix insist on port names like 'ttxS[n]'.
#' # 
#' # Second: Set up a terminal program (like HTerm or gtkterm) and listen to 
#' #         COM port 'CNCB0' (or what ever you have installed)
#' #         or (for unix only) 'cat /dev/tnt1' will output tnt1 to the console.
#' 
#' \dontrun{
#' 
#' # Now configure one of the COM ports with appropriate connection properties.
#' con <- serialConnection(port = "CNCA0", # or "tnt0" on linux
#'                         mode = "115200,n,8,1",
#'                         newline = 1,
#'                         translation = "crlf")
#' 
#' # open the serial interface,
#' open(con)
#' 
#' # write some stuff,
#' write(con, "Hello World!")
#' 
#' # read, in case something came in,
#' read(con)
#' 
#' # and close the connection.
#' close(con)
#' }
#' @concept RS232 serial
#' @concept RS485 serial
#' @concept RS422 serial
#' @concept USB serial
#' 
#' @docType package
#' @name serial
NULL

R/utilities.R
# Utilities

#' Find serial ports in use.
#' 
#' This function checks for serial ports in use on the host device.
#' 
#' @usage FindDevices()
#' 
#' @return A vector of strings containing the names of all active devices, or 
#' `character(0)` if no device is found.
#' @examples
#' \dontrun{
#' # Use the first active device.
#' available_devices <- FindDevices()
#' 
#' con <- serialConnection(available_devices[1])
#' }
#' @importFrom utils file_test
#' @export
FindDevices <- function() {
  os <- get_os()
  if (identical(os, "linux")) {
    if (file_test("-d", "/dev/serial/by-id")) { # dir only exists when device connected
      # check symbolic links
      cmmd <- 'ls -l /dev/serial/by-id | grep "\\->"'
      dev_info <- system(cmmd, intern = TRUE)
      regmatches(dev_info, regexpr('(?:tty).*', text = dev_info))
    } else {
      character(0)
    }
  } 
  
  else if (identical(os, "win")) {
    cmmd <- 'REG QUERY HKEY_LOCAL_MACHINE\\HARDWARE\\DEVICEMAP\\SERIALCOMM'
    dev_info <- system(cmmd, intern = TRUE)
    regmatches(dev_info, regexpr('(?:COM)[0-9]{1,3}', text = dev_info))
  } 
  
  else if (identical(os, "osx")) {
    dev_info <- system('ls /dev/tty.*', intern = TRUE)
    gsub("/dev/", "", dev_info)
  } 
  
  else {
    stop("Unidentified platform.")
  }
}

# Try to find out the host operating system
get_os <- function() {
  ifelse(identical(.Platform$OS.type, 'windows'), 'win', 
               ifelse(grepl("darwin", R.version$os), 'osx', 'linux'))
}


R/serial.R
#' Sets up the interface parameters.
#' 
#' @param port  the COM port name, should also work in Linux. Also, virtual COMs are 
#'              supported, meaning USB should work too.
#' @param mode  the communication mode "\code{<BAUD>, <PARITY>, <DATABITS>, <STOPBITS>}".
#' \describe{
#'    \item{\code{BAUD}}{sets the baud rate (bits per second).}
#'    \item{\code{PARITY}}{\emph{n, o, e, m, s} stands for "none", "odd", "even", "mark" and "space".}
#'    \item{\code{DATABITS}}{integer number of data bits. The value can range from 5 to 8.}
#'    \item{\code{STOPBITS}}{integer number of stop bits. This can be "1" or "2".}
#'        }       
#'              
#' @param buffering "\code{none}", for RS232 serial interface only; other modes don't work in this case.
#' @param newline \code{<BOOL>}, whether a new transmission starts with a newline or not.
#'                \describe{
#'                  \item{\code{TRUE} or 1}{send newline-char according to \code{<translation>} before transmitting.}
#'                  \item{\code{FALSE} or 0}{no newline.}
#'                          }
#' @param handshake determines the type of handshaking the communication.
#'                  \describe{
#'                    \item{"\code{none}"}{no handshake is done.}
#'                    \item{"\code{rtscts}"}{hardware handshake is enabled.}
#'                    \item{"\code{xonxoff}"}{software handshake via extra characters is enabled.}
#'                    }
#' 
#' @param eof \code{<CHAR>}, termination char of the datastream. It only makes sense
#'        if \code{<translation>} is "binary" and the stream is a file.
#' @param translation each transmitted string is terminated by the transmission
#'       character. This could be "lf", "cr", "crlf", or "binary".
#' @return An object of the class "\code{serialConnection}" is returned.
#' @seealso \code{\link{serial}} \code{\link{open}} \code{\link{close}}
#' @export
serialConnection <- function(port = "com1", mode = "115200,n,8,1", 
                             buffering = "none", newline = 0, eof = "",
                             translation = "lf", handshake= "none") {
  
  obj <- as.list(environment())
  class(obj) <- "serialConnection"
  obj
}

#' Function to initialize an serial interface.
#' 
#' This function initializes the serial interface and opens it for later usage. 
#' 
#' @name open
#' @method open serialConnection
#' @aliases open open.serialConnection
#' 
#' @param con Object of class \code{serialConnection}.
#' @param ... Currently ignored.
#' @seealso \code{\link{serialConnection}} \code{\link{close}}
#' @examples
#' \dontrun{
#' con <- serialConnection(port = 'com5')
#' open(con)
#' }
#' @importFrom tcltk .Tcl
#' @export
open.serialConnection <- function(con, ...) {
  ## set platform-dependent path
  os_path <- switch(.Platform$OS.type,
                    windows = "//./",
                    unix = "/dev/")
  ## set connection and variables
  .Tcl(paste("set sdev_", con$port," [open ", os_path, con$port, " r+]", sep = ""))
  ## '//./' defines the Windows path for the COM ports
  ## only in this way it is possible to use virtual ports as well
  ## the unix COM ports are located in '/dev/'
  
  ## set up configuration
  eof <- ifelse(con$eof == "", "", paste(" -eofchar ", con$eof, sep = ""))
  
  .Tcl(paste("fconfigure $sdev_",con$port,
              " -mode ", con$mode,
              " -buffering ", con$buffering,
              " -blocking 0", eof,
              " -translation ", con$translation,
              " -handshake ", con$handshake,
              " -timeout ", 5000, sep = ""))
  invisible(NULL)
  ## it seems that -eofchar doesn't work
  ## "buffering none" is recommended, other setings doesn't work to send 
}

#' Function to close a serial interface.
#' 
#' This function closes the corresponding connection.
#' 
#' @name close
#' @method close serialConnection
#' @aliases close close.serialConnection
#' @param con Object of class \code{serialConnection}.
#' @param ... Currently ignored.
#' 
#' @examples
#' \dontrun{
#' close(con)
#' }
#' 
#' @seealso \code{\link{serialConnection}} \code{\link{open}}
#' @importFrom tcltk .Tcl
#' @export
close.serialConnection <- function(con, ...) {
  .Tcl(paste("close $sdev_", con$port, sep = ""))
  invisible(NULL)
}


#' Reads from the serial interface.
#' 
#' This function reads from the serial interface as long as the buffer is not
#' empty. The read takes place per byte.
#' 
#' @name read
#' @param con Object of class \code{serialConnection}
#' @param ... Currently ignored.
#' 
#' @aliases read read.serialConnection
#' 
#' @return The result is a string, which can be converted to raw as necessary.
#' @seealso \code{\link{serial}} \code{\link{write}}
#' @examples
#'# See `help(serial)` for a complete example.
#'\dontrun{
#'data_in <- read(con)
#'}
#' @importFrom tcltk .Tcl tclvalue
#' @export 
read.serialConnection <- function(con, ...) {
  res <- ""
  while(1) {
    tmp <- tclvalue(.Tcl(paste("gets $sdev_", con$port, sep = "")))
    if (tmp == "") break
    res <- paste(res, tmp, sep = "")
  }
  res
}


#' Writes data to the serial interface. 
#' 
#' @name write
#' @param x Object of class \code{serialConnection}
#' @param dat Data string to write on the serial interface. This must be a character string.
#' @param ... Currently ignored.
#' @aliases write write.serialConnection
#' @seealso \code{\link{serial}} \code{\link{read}}
#' @examples
#'# See `help(serial)` for a complete example.
#'  
#'\dontrun{
#'write(con, "Hello World!")
#'write(con, 2) # error
#'}
#' @importFrom methods is
#' @importFrom tcltk .Tcl
#' @export 
write.serialConnection <- function(x, dat, ...) {
  
  if (!is(dat, 'character')) {
    stop("dat is not a character.")
  }
  
  nl <- ifelse(x$newline, "", "-nonewline ")
  .Tcl(paste("puts ", nl, "$sdev_", x$port, " \"", dat, "\"", sep = ""))
  #   ..," \"", dat,"\"",.. -> quotes dat in TCL String
  #   with out quoting space and control characters this will fail 
  invisible(NULL)
}


# Deprecated

#' #' @export
#' write.serialConnection <- function(con, dat, ...) UseMethod("write")
#' 
#' #' @export
#' read.serialConnection <- function(con, ...) UseMethod("read")